worker/modelcache/worker_test.go

```diff
@@ -588,6 +615,60 @@ func (s *WorkerSuite) TestWatcherErrorCacheMarkSweep(c *gc.C) {
 	c.Check(cachedApp, gc.NotNil)
 }
 
+func (s *WorkerSuite) TestWatcherErrorRestartBackoff(c *gc.C) {
+	clk := testclock.NewClock(time.Now())
+
+	s.config.WatcherRestartDelayMin = 2 * time.Second
+	s.config.WatcherRestartDelayMax = time.Minute
+	s.config.Clock = clk
+
+	// 7 times through the loop will double the timeout (starting at 2s) until
+	// we get two times though with the max delay (1m).
+	maxErrors := 7
+
+	var errCount int
+	s.config.WatcherFactory = func() modelcache.BackingWatcher {
+		return testingMultiwatcher{
+			Multiwatcher: s.StatePool.SystemState().WatchAllModels(s.StatePool),
+			manipulate: func(deltas []multiwatcher.Delta) ([]multiwatcher.Delta, error) {
+				if errCount < maxErrors {
+					errCount++
+					return nil, errors.New("boom")
+				}
+				return deltas, nil
+			},
+		}
+	}
+
+	changes := s.captureEvents(c, cachetest.ModelEvents, cachetest.ApplicationEvents)
+	_ = s.start(c)
+	s.State.StartSync()
+
+	// Until the watcher returns without error, advance the clock by the exact
+	// duration we expect the restart delay to be based on our initial config.
+	for i := 1; i <= maxErrors; i++ {
+		delay := time.Duration(math.Pow(2, float64(i))) * time.Second
```

Comment:
I guess since we're using powers of two, we don't have to worry about float rounding funny causing this test to break?